src/main.cpp

```diff
@@ -3446,131 +3500,104 @@ bool ActivateBestChain(CValidationState &state, const CChainParams &chainparams,
 {
     CBlockIndex *pindexMostWork = NULL;
     LOCK(cs_main);
+
+    bool fOneDone = false;
     do
     {
         boost::this_thread::interruption_point();
         if (ShutdownRequested())
-            break;
+            return false;
 
-        CBlockIndex *pindexNewTip = NULL;
-        const CBlockIndex *pindexFork;
-        {
-            CBlockIndex *pindexOldTip = chainActive.Tip();
-            pindexMostWork = FindMostWorkChain();
+        CBlockIndex *pindexOldTip = chainActive.Tip();
+        pindexMostWork = FindMostWorkChain();
 
-            // This is needed for PV because FindMostWorkChain does not necessarily return the block with the lowest
-            // nSequenceId
-            if (fParallel)
+
+        // This is needed for PV because FindMostWorkChain does not necessarily return the block with the lowest
+        // nSequenceId
+        if (fParallel && pblock)
+        {
+            std::set<CBlockIndex *, CBlockIndexWorkComparator>::reverse_iterator it = setBlockIndexCandidates.rbegin();
+            while (it != setBlockIndexCandidates.rend())
             {
-                std::set<CBlockIndex *, CBlockIndexWorkComparator>::reverse_iterator it =
-                    setBlockIndexCandidates.rbegin();
-                while (it != setBlockIndexCandidates.rend())
-                {
-                    if ((*it)->nChainWork == pindexMostWork->nChainWork)
-                        if ((*it)->nSequenceId < pindexMostWork->nSequenceId)
-                            pindexMostWork = *it;
-                    it++;
-                }
+                if ((*it)->nChainWork == pindexMostWork->nChainWork)
+                    if ((*it)->nSequenceId < pindexMostWork->nSequenceId)
+                        pindexMostWork = *it;
+                it++;
             }
+        }
 
-            // Whether we have anything to do at all.
-            if (pindexMostWork == NULL)
+        // Whether we have anything to do at all.
+        if (pindexMostWork == NULL)
```

Comment:
In the new code you use pindexMostWork before checking it for NULL.  Here, someone clearly thought that NULL was possible.  Perhaps this null check and return should be moved above the if (fParallel) {}?